core/chains/evm/client/node_lifecycle_test.go

```diff
@@ -315,10 +316,194 @@ func TestUnit_NodeLifecycle_aliveLoop(t *testing.T) {
+		n := iN.(*node)
+		n.nLiveNodes = func() (count int, blockNumber int64, totalDifficulty *utils.Big) {
+			return 2, highestHead.Load(), utils.NewBigI(0)
+		}
+
+		dial(t, n)
+		defer n.Close()
+
+		n.wg.Add(1)
+		go n.aliveLoop()
+
+		testutils.WaitWithTimeout(t, chSubbed, "timed out waiting for initial subscription for InSync")
+
+		// ensure alive up to stall
+		testutils.AssertEventually(t, func() bool {
+			state, num, _ := n.StateAndLatest()
+			if num < stall {
+				require.Equal(t, NodeStateAlive, state)
+			}
+			return num == stall
+		})
+
+		testutils.AssertEventually(t, func() bool {
+			state, num, _ := n.StateAndLatest()
+			return state == NodeStateOutOfSync && num == stall
+		})
+		assert.GreaterOrEqual(t, highestHead.Load(), int64(stall+cfg.SyncThreshold))
+
+		// Otherwise, there may be data race on dial() vs Close() (accessing ws.rpc)
+		testutils.WaitWithTimeout(t, chSubbed, "timed out waiting for initial subscription for OutOfSync")
+	})
+
+	t.Run("when behind but SyncThreshold=0, stay alive", func(t *testing.T) {
+		cfg := TestNodeConfig{SyncThreshold: 0, PollFailureThreshold: 2, PollInterval: 100 * time.Millisecond, SelectionMode: NodeSelectionMode_HighestHead}
+		chSubbed := make(chan struct{}, 1)
+		var highestHead atomic.Int64
+		const stall = 10
+		s := testutils.NewWSServer(t, testutils.FixtureChainID,
+			func(method string, params gjson.Result) (resp testutils.JSONRPCResponse) {
+				switch method {
+				case "eth_subscribe":
+					chSubbed <- struct{}{}
+					resp.Result = `"0x00"`
+					resp.Notify = makeHeadResult(int(highestHead.Load()))
+					return
+				case "eth_unsubscribe":
+					resp.Result = "true"
+					return
+				case "web3_clientVersion":
+					resp.Result = `"test client version 2"`
+					// always tick each poll, but only signal back up to stall
+					if n := highestHead.Inc(); n <= stall {
+						resp.Notify = makeHeadResult(int(n))
+					}
+					return
+				default:
+					t.Errorf("unexpected RPC method: %s", method)
+				}
+				return
+			})
+
+		iN := NewNode(cfg, logger.TestLogger(t), *s.WSURL(), nil, "test node", 42, testutils.FixtureChainID)
+		n := iN.(*node)
+		n.nLiveNodes = func() (count int, blockNumber int64, totalDifficulty *utils.Big) {
+			return 2, highestHead.Load(), utils.NewBigI(0)
+		}
+
+		dial(t, n)
+		defer n.Close()
+
+		n.wg.Add(1)
+		go n.aliveLoop()
+
+		testutils.WaitWithTimeout(t, chSubbed, "timed out waiting for initial subscription for InSync")
+
+		// ensure alive up to stall
+		testutils.AssertEventually(t, func() bool {
+			state, num, _ := n.StateAndLatest()
+			require.Equal(t, NodeStateAlive, state)
+			return num == stall
+		})
+
+		assert.Equal(t, NodeStateAlive, n.state)
+		assert.GreaterOrEqual(t, highestHead.Load(), int64(stall+cfg.SyncThreshold))
+	})
+
+	t.Run("when behind more than SyncThreshold but we are the last live node, forcibly stays alive", func(t *testing.T) {
```

Comment:
Presumably the case of `"when ahead of SyncThreshold, stays alive/does nothing"` is already implicitly covered by other tests?

Reply:
Partially, but we also begin in that state for most of these tests too, before inducing the problem at the `stall` block.